### DISCUSSION TITLE
Ahead-of-time compilation for flax `TrainState`

### DISCUSSION THREAD

**Discussion by mar-muel** [google/flax#3100] [Q&A] [answered]
Hi there! I'm trying to set up [ahead-of-time compilation](https://jax.readthedocs.io/en/latest/aot.html) for my training step function.

Currently my `train_step` function takes in a [flax training state](https://flax.readthedocs.io/en/latest/_modules/flax/training/train_state.html) as an argument. After I `lower()` and `compile()` this function, I'm trying to call the compiled function. However, I'm getting the following error:
```
TypeError: function compiled for PyTreeDef(((CustomNode(TrainState[
[...long list of arguments...]
called with PyTreeDef(((CustomNode(TrainState[(<bound method
[...long list of arguments...]
```
Basically the error tells me that I called the compiled version with a different instance of TrainState (different memory location).

So I gather from this that I cannot use AOT-compilation with flax due to the train state having a dynamic nature that requires compilation every time it is newly created? Thanks for any insight into this!

**Comment by cgarciae** [comment]
Hey @mar-muel, can you send a minimal reproducible example? I tried with the following code and it worked fine:

```python
import flax.linen as nn
import jax.numpy as jnp
import jax
from flax.training.train_state import TrainState
import optax

# data
x = jax.random.uniform(jax.random.PRNGKey(0), [32, 784])
y = jax.random.randint(jax.random.PRNGKey(1), [32], 0, 10)

# create state
module = nn.Dense(10)
params = module.init(jax.random.PRNGKey(0), x)['params']
state = TrainState.create(apply_fn=module.apply, params=params, tx=optax.adam(0.01))

@jax.jit
def train_step(state: TrainState, batch):
    x, y = batch
    
    def loss_fn(params):
        logits = module.apply({'params': params}, x)
        return jnp.mean(optax.softmax_cross_entropy_with_integer_labels(logits, y))

    loss, grad = jax.value_and_grad(loss_fn)(state.params)
    state = state.apply_gradients(grads=grad)
    return state, loss

# compile
c_train_step = train_step.lower(state, (x, y)).compile()

# test
state, loss = c_train_step(state, (x, y))

print(jax.tree_map(lambda x: x.shape, state.params))
print(loss)
```

**Comment by mar-muel** [comment]
Hi @cgarciae - Thanks for getting back to me!

The issue only appears when you pass a new state to the compiled function:

```python
import flax.linen as nn
import jax.numpy as jnp
import jax
from flax.training.train_state import TrainState
import optax

# data
x = jax.random.uniform(jax.random.PRNGKey(0), [32, 784])
y = jax.random.randint(jax.random.PRNGKey(1), [32], 0, 10)

# create state
module = nn.Dense(10)
params = module.init(jax.random.PRNGKey(0), x)['params']
state = TrainState.create(apply_fn=module.apply, params=params, tx=optax.adam(0.01))

@jax.jit
def train_step(state: TrainState, batch):
    x, y = batch

    def loss_fn(params):
        logits = module.apply({'params': params}, x)
        return jnp.mean(optax.softmax_cross_entropy_with_integer_labels(logits, y))

    loss, grad = jax.value_and_grad(loss_fn)(state.params)
    state = state.apply_gradients(grads=grad)
    return state, loss

# compile
c_train_step = train_step.lower(state, (x, y)).compile()

# at some point later in the code we create a new training state...
params = module.init(jax.random.PRNGKey(0), x)['params']
state = TrainState.create(apply_fn=module.apply, params=params, tx=optax.adam(0.01))

# run with new state
state, loss = c_train_step(state, (x, y))

print(jax.tree_map(lambda x: x.shape, state.params))
print(loss)
```

Gives
```
Traceback (most recent call last):
  File "test_flax_aot.py", line 36, in <module>
    state, loss = c_train_step(state, (x, y))
  File "/home/martin/miniconda/envs/testenv/lib/python3.8/site-packages/jax/_src/stages.py", line 517, in __call__
    return self._cpp_call(*args, **kwargs)
  File "/home/martin/miniconda/envs/testenv/lib/python3.8/site-packages/jax/interpreters/pxla.py", line 3589, in aot_cache_miss
    outs, out_flat, args_flat = stages.Compiled.call(params, *args, **kwargs)
  File "/home/martin/miniconda/envs/testenv/lib/python3.8/site-packages/jax/_src/stages.py", line 491, in call
    raise TypeError(
TypeError: function compiled for PyTreeDef(((CustomNode(TrainState[(<bound method Module.apply of Dense(
    # attributes
    features = 10
    use_bias = True
    dtype = None
    param_dtype = float32
    precision = None
    kernel_init = init
    bias_init = zeros
)>, GradientTransformation(init=<function chain.<locals>.init_fn at 0x7f984b01e310>, update=<function chain.<locals>.update_fn at 0x7f984b01e3a0>))], [*, CustomNode(FrozenDict[()], [{'bias': *, 'kernel': *}]), (CustomNode(namedtuple[ScaleByAdamState], [*, CustomNode(FrozenDict[()], [{'bias': *, 'kernel': *}]), CustomNode(FrozenDict[()], [{'bias': *, 'kernel': *}])]), CustomNode(namedtuple[EmptyState], []))]), (*, *)), {})), called with PyTreeDef(((CustomNode(TrainState[(<bound method Module.apply of Dense(
    # attributes
    features = 10
    use_bias = True
    dtype = None
    param_dtype = float32
    precision = None
    kernel_init = init
    bias_init = zeros
)>, GradientTransformation(init=<function chain.<locals>.init_fn at 0x7f9848b211f0>, update=<function chain.<locals>.update_fn at 0x7f9848b21790>))], [*, CustomNode(FrozenDict[()], [{'bias': *, 'kernel': *}]), (CustomNode(namedtuple[ScaleByAdamState], [*, CustomNode(FrozenDict[()], [{'bias': *, 'kernel': *}]), CustomNode(FrozenDict[()], [{'bias': *, 'kernel': *}])]), CustomNode(namedtuple[EmptyState], []))]), (*, *)), {}))
```

My (currently somewhat outdated) versions:
```
jax==0.4.2
jaxlib==0.4.2
flax==0.6.3
```

**Comment by cgarciae** [comment, marked as answer]
@mar-muel I see. You can make it work by reusing the same `optax` optimizer instance so the hash for the `state` doesn't change:

```python
from pyexpat import model
import flax.linen as nn
import jax.numpy as jnp
import jax
from flax.training.train_state import TrainState
import optax

# data
x = jax.random.uniform(jax.random.PRNGKey(0), [32, 784])
y = jax.random.randint(jax.random.PRNGKey(1), [32], 0, 10)

# create state
module = nn.Dense(10)
tx = optax.adam(0.01) # <== reuse tx
params = module.init(jax.random.PRNGKey(0), x)['params']
state = TrainState.create(apply_fn=module.apply, params=params, tx=tx)

@jax.jit
def train_step(state: TrainState, batch):
    x, y = batch
    
    def loss_fn(params):
        logits = module.apply({'params': params}, x)
        return jnp.mean(optax.softmax_cross_entropy_with_integer_labels(logits, y))

    loss, grad = jax.value_and_grad(loss_fn)(state.params)
    state = state.apply_gradients(grads=grad)
    return state, loss

# compile
c_train_step = train_step.lower(state, (x, y)).compile()

# test
state = TrainState.create(apply_fn=module.apply, params=params, tx=tx)
state, loss = c_train_step(state, (x, y))

print(jax.tree_map(lambda x: x.shape, state.params))
print(loss)
```

**Reply by mar-muel** [reply]
Alright yeah that's it - Thanks 🙏
